pubsublite/topiccreator.go

```diff
@@ -0,0 +1,164 @@
+// specific language governing permissions and limitations
+// under the License.
+
+package pubsublite
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"cloud.google.com/go/pubsublite"
+	"github.com/googleapis/gax-go/v2/apierror"
+	"go.uber.org/zap"
+
+	apmqueue "github.com/elastic/apm-queue"
+)
+
+// TopicCreatorConfig holds configuration for managing GCP Pub/Sub Lite topics.
+type TopicCreatorConfig struct {
+	// Reservation holds the unqualified ID of the reservation with
+	// which topics will be associated. This will be combined with the
+	// project ID and region ID to form the qualified reservation name.
+	Reservation string
+
+	// PartitionCount is the number of partitions to assign to newly
+	// created topics.
+	//
+	// Must be greater than zero.
+	PartitionCount int
+
+	// PublishCapacityMiBPerSec defines the publish throughput capacity
+	// per partition in MiB/s. Must be >= 4 and <= 16.
+	PublishCapacityMiBPerSec int
+
+	// SubscribeCapacityMiBPerSec defines the subscribe throughput capacity
+	// per partition in MiB/s. Must be >= 4 and <= 32.
+	SubscribeCapacityMiBPerSec int
+
+	// PerPartitionBytes holds the provisioned storage, in bytes, per partition.
+	//
+	// If the number of bytes stored in any of the topic's partitions grows beyond
+	// this value, older messages will be dropped to make room for newer ones,
+	// regardless of the value of `RetentionDuration`. Must be >= 30 GiB.
+	PerPartitionBytes int64
+
+	// RetentionDuration indicates how long messages are retained. Must be > 0.
+	RetentionDuration time.Duration
+}
+
+// Validate checks that cfg is valid, and returns an error otherwise.
+func (cfg TopicCreatorConfig) Validate() error {
+	var errs []error
+	if cfg.Reservation == "" {
+		errs = append(errs, errors.New("pubsublite: reservation must be set"))
+	}
+	if cfg.PartitionCount <= 0 {
+		errs = append(errs, errors.New("pubsublite: partition count must be greater than zero"))
+	}
+	if cfg.PublishCapacityMiBPerSec < 4 || cfg.PublishCapacityMiBPerSec > 16 {
+		errs = append(errs, errors.New("pubsublite: publish capacity must between 4 and 16, inclusive"))
+	}
+	if cfg.SubscribeCapacityMiBPerSec < 4 || cfg.SubscribeCapacityMiBPerSec > 32 {
+		errs = append(errs, errors.New("pubsublite: subscribe capacity must between 4 and 32, inclusive"))
+	}
+	if cfg.PerPartitionBytes < 30*1024*1024*1024 {
+		errs = append(errs, errors.New("pubsublite: per-partition bytes must be at least 30GiB"))
+	}
+	if cfg.RetentionDuration <= 0 {
+		errs = append(errs, errors.New("pubsublite: retention duration must be greater than zero"))
+	}
+	return errors.Join(errs...)
+}
+
+var _ apmqueue.TopicCreator = (*TopicCreator)(nil)
+
+// TopicCreator creates GCP Pub/Sub topics.
+type TopicCreator struct {
+	m   *Manager
+	cfg TopicCreatorConfig
+}
+
+// NewTopicCreator returns a new TopicCreator with the given config.
+func (m *Manager) NewTopicCreator(cfg TopicCreatorConfig) (*TopicCreator, error) {
```

Comment:
For kafka, the topic creation is held in the manager, while pubsublite has its own independent struct. What is the reason for not putting this logic within the `pubsublite.Manager` struct?

Reply:
This is so we can have a common CreateTopics signature/interface.

pubsublite has this concept of a throughput reservation, which kafka doesn't have. The pubsublite.Manager is not bound to a specific reservation (it can't be, it manages them); the TopicCreator is bound to one, so its CreateTopics method can create topics attached to that reservation.